lib/tram.ex

```diff
@@ -0,0 +1,79 @@
+defmodule Tram do
+  use GenServer
+
+  # Client
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, :ok)
+  end
+
+  def stop() do
+    GenServer.call(__MODULE__, :stop)
+  end
+
+  def go() do
+    GenServer.call(__MODULE__, :stop)
+  end
+
+  def open_doors() do
+    GenServer.call(__MODULE__, :open_doors)
+  end
+
+  def close_doors() do
+    GenServer.call(__MODULE__, :open_doors)
+  end
+
+  def end_shift() do
+    GenServer.cast(__MODULE__, :end_shift)
+  end
+
+  def error_message() do
+    {:error, "Operation is denied"}
+  end
+
+  # Server (callbacks)
+
+  @impl true
+  def init(:ok) do
+    {:ok, :start_shift}
+  end
+
+  @impl true
+  def handle_call(:stop, _from, state) do
+    case state do
+      :start_shift -> {:reply, :ok, :stop}
+      _invalid_state -> {:reply, :invalid_transition, state}
+    end
+  end
```

Comment:
Отвыкайте от вложенных проверок.
```suggestion
  def handle_call(:stop, _from, :start_shift), do: {:reply, :ok, :stop}
  def handle_call(:stop, _from, _invalid_state), do: {:reply, :invalid_transition, state}
```